lib/model/unittest/CModelTestFixtureBase.cc

```diff
@@ -0,0 +1,310 @@
+                                                     ml::model::CModelFactory::TDataGathererPtr& gatherer,
+                                                     ml::model::CAnomalyDetectorModel& model,
+                                                     TAnomalyVec& orderedAnomalies) {
+
+    TAnomalyAccumulator anomalies(numAnomalies);
+
+    for (std::size_t i = 0u, bucket = 0u; i < messages.size(); ++i) {
+        if (messages[i].s_Time >= startTime + bucketLength) {
+            LOG_DEBUG(<< "Updating and testing bucket = [" << startTime << ","
+                      << startTime + bucketLength << ")");
+
+            model.sample(startTime, startTime + bucketLength, m_ResourceMonitor);
+
+            ml::model::CPartitioningFields partitioningFields(EMPTY_STRING, EMPTY_STRING);
+            ml::model::SAnnotatedProbability annotatedProbability;
+            for (std::size_t pid = 0u; pid < gatherer->numberActivePeople(); ++pid) {
+                model.computeProbability(pid, startTime, startTime + bucketLength,
+                                         partitioningFields, 2, annotatedProbability);
+
+                std::string person = model.personName(pid);
+                TDoubleStrPrVec attributes;
+                for (const auto& probability : annotatedProbability.s_AttributeProbabilities) {
+                    attributes.emplace_back(probability.s_Probability,
+                                            *probability.s_Attribute);
+                }
+                LOG_DEBUG(<< "Adding anomaly " << pid);
+                anomalies.add({annotatedProbability.s_Probability,
+                               {bucket, person, attributes}});
+            }
+
+            startTime += bucketLength;
+            ++bucket;
+        }
+
+        this->addArrival(messages[i], gatherer);
+    }
+
+    anomalies.sort();
+    LOG_DEBUG(<< "Anomalies = " << anomalies.print());
+
+    for (std::size_t i = 0u; i < anomalies.count(); ++i) {
+        orderedAnomalies.push_back(anomalies[i].second);
+    }
+
+    std::sort(orderedAnomalies.begin(), orderedAnomalies.end());
+
+    LOG_DEBUG(<< "orderedAnomalies = "
+              << ml::core::CContainerPrinter::print(orderedAnomalies));
+}
+
+void CModelTestFixtureBase::generateAndCompareKey(const ml::model::function_t::TFunctionVec& countFunctions,
+                                                  const std::string& fieldName,
+                                                  const std::string& overFieldName,
+                                                  TKeyCompareFunc keyCompare) {
+    TBoolVec useNull{true, false};
+    TStrVec byFields{"", "by"};
+    TStrVec partitionFields{"", "partition"};
+
+    ml::model::CAnomalyDetectorModelConfig config =
+        ml::model::CAnomalyDetectorModelConfig::defaultConfig();
+
+    int detectorIndex{0};
+    for (const auto& countFunction : countFunctions) {
+        for (bool usingNull : useNull) {
```

Comment:
Probably just moved here, but this can be more concise:
```suggestion
        for (bool usingNull : {true, false}) {
```
then delete `useNull`. Similar for others below.